apps/web/src/app/[locale]/challenge/_components/comments/comments.hooks.tsx

```diff
@@ -0,0 +1,296 @@
+import { useReducer } from 'react';
+
+import { commentErrors, sortKeys } from './comments.constants';
+import { getPaginatedComments } from './getCommentRouteData';
+import type { CommentRoot } from '@repo/db/types';
+import {
+  addComment as addCommentAction,
+  deleteComment as deleteCommentAction,
+  replyComment,
+  updateComment as updateCommentAction,
+} from './comment.action';
+import { toast } from '@repo/ui/components/use-toast';
+
+const getRootQueryKey = (rootId: number, type: CommentRoot) =>
+  `${type.toLowerCase()}-${rootId}-comments`;
+
+export type SortItem = (typeof sortKeys)[number];
+
+interface CommentsMeta {
+  page: number;
+  sort: SortItem;
+}
+
+interface DefaultCommentsProps {
+  rootId: number;
+  type: CommentRoot;
+}
+
+interface UseCommentsProps extends DefaultCommentsProps {
+  initialPage?: number;
+}
+
+export function useComments({ type, rootId, initialPage }: UseCommentsProps) {
+  const queryClient = useQueryClient();
+  const [meta, updateMeta] = useReducer(
+    (state: CommentsMeta, action: Partial<CommentsMeta>) => ({ ...state, ...action }),
+    {
+      page: initialPage ?? 1,
+      sort: sortKeys[0],
+    },
+  );
+
+  const getQueryKey = ({ sort, page }: { sort: string; page: number }) => [
+    getRootQueryKey(rootId, type),
+    sort,
+    page,
+  ];
+
+  const { status, data } = useQuery({
+    queryKey: getQueryKey({ sort: meta.sort.value, page: meta.page }),
+    queryFn: () => {
+      return getPaginatedComments({
+        rootId,
+        page: meta.page,
+        rootType: type,
+        sortKey: meta.sort.key,
+        sortOrder: meta.sort.order,
+      });
+    },
+    placeholderData: keepPreviousData,
+    staleTime: 60000, // one minute
+    refetchOnWindowFocus: false,
+  });
+
+  const changePage = (page: number) => {
+    updateMeta({ page });
+  };
+
+  const changeSorting = (sort: string) => {
+    updateMeta({ sort: sortKeys.find((key) => key.value === sort) ?? sortKeys[0], page: 1 });
+  };
+
+  const deleteComment = async (commentId: number) => {
+    try {
+      const res = await deleteCommentAction(commentId);
+      if (res === 'unauthorized') {
+        toast(commentErrors.unauthorized);
+      } else if (res === 'invalid_comment') {
+        toast(commentErrors.invalidId);
+      } else {
+        toast({
+          title: 'Comment Deleted',
+          variant: 'success',
+          description: 'The comment was successfully deleted.',
+        });
+      }
+    } catch (e) {
+      toast({
+        ...commentErrors.unexpected,
+        variant: 'destructive',
+      });
+    } finally {
+      // If the last comment on the page was deleted, we need to go back a page
+      const newPage = data?.comments.length === 1 ? meta.page - 1 : meta.page;
+      changePage(newPage);
+      queryClient.invalidateQueries({
+        queryKey: getQueryKey({ sort: meta.sort.value, page: newPage }),
+      });
+    }
+  };
+
+  const addComment = async (text: string) => {
```

Comment:
should these be using `useCallback` looks like youre passing some of these down to components

Reply:
IMO, not necessary since all actions will be re-invalidating queries, which will trigger refetches, and deps will change in those `useCallbacks`. 
I haven't noticed any performance issues so far, so let's not try to fix it prematurely